text/2020-03-11-versioned-kv.md

```diff
@@ -0,0 +1,395 @@
+  - Put/BatchPut -  Increases the new version value of a key, where version is the current timestamp by default
+  - Delete/BatchDelete - Deletes all versions of a specified key (`key:cur_ts -> delete`)
+  - DeleteRange - Deletes all versions of key in a range using rocksdb DeleteRange
+
+The above fetch operations(get/scan) can use `rocksdb prefix_extractor` to improve the performance of iterator seek. `key: ts-> delete` indicates that all data before the `ts` has been deleted.
+
+### Version recycling
+
+The design of the version recycling will affect cluster performance and the functional designs of incremental backup and CDC. There are two major approaches: 
+  - using multi-version Garbage Collection (GC) similar to TxnKV, and periodically starting the gc worker in the background to delete redundant versions of data;
+  - using compaction filter via rocksdb compaction to delete redundant versions. 
+
+The GC method consumes CPU or storage resources, but the version recycling is accurate and efficient; The compaction filter method does not take up CPU or storage resources. However, compaction only has a local applicable scope, so multi-version recycling efficiency is low.
+
+This proposal adopts a combination of the above two methods, with compaction as the major mechanism and GC as the supplement. The default maximum number of versions is 1, which can default to the compaction filter recycling version, so recycling efficiency is moderate. If the maximum number of versions is more than 1 or more efficient version recycling is required, you can dynamically configure GC to enable it.
+
+The compaction filter is customized based on the maximum number of versions, `MaxVerNum`, and delete tombstone based on key instead of key: `ts`. So only one `MaxVerNum` can be set for one `cf`.
+
+### Data increment
+
+The `ScanNumWithSV` rpc interface is used to get all data increments  of the entire cluster or within a specified range [start_ts, cur_ts). When this interface is called, the snapshot of  the entire cluster or the current cur_ts is obtained before scanning.
+
+VerKV integrated to tidb/br and tidb/ticdc can better facilitate the implementation of full/incremental backup and CDC services.  
+
+### Data replication
+
+- A-> B - asynchronous one-way replication; Cluster B accesses the CDC service of cluster A to pull or push incremental data of each period.
+- A <-> B -  asynchronous two-way replication, which is equivalent to two one-way replication processes. Write conflicts can be resolved according to ts.
+
+### Others
+
+Rocksdb User Timestamp（WIP）can also be considered to improve the engine efficiency. 
+
+## Drawbacks
+
+This VerKV adds version information, so it would decrease the performance of data fetch interface. And the data version recycling would also add pressure to  system.
+
+## Appendix
+
+### Proto3 definition
```

Comment:
I suggest to remove the definitions. They are subject to change and becomes misleading/useless in the end.

Reply:
done